client/Client.hs

```diff
@@ -1,30 +1,368 @@
+
+unsub :: TVar Client -> SID -> IO ()
+unsub client sid = atomically $ unsub' client sid
+
+pub :: TVar Client -> [PubOptions -> PubOptions] -> IO ()
+pub client options = atomically $ pub' client options
+
+pong :: TVar Client -> IO ()
+pong = atomically . pong'
+
+ping :: TVar Client -> IO ()
+ping = atomically . ping'
+
+stop :: TVar Client -> IO ()
+stop = atomically . stop'
+
+defaultConn :: String -> Int -> IO Handle
+defaultConn host port = do
+  (sock, _) <- TCP.connectSock host $ show port
+  NS.setSocketOption sock NS.NoDelay 1
+  NS.setSocketOption sock NS.Cork 0
+  NS.setSocketOption sock NS.RecvBuffer 1
+  NS.setSocketOption sock NS.SendBuffer 1
+  handle <- NS.socketToHandle sock ReadWriteMode
+  hSetBuffering handle NoBuffering
+  return handle
+
+instance NatsConn Handle where
+  send = BS.hPut
+  recv = BS.hGetNonBlocking
+  close = hClose
+
+withNats :: NatsConn c => [ClientOption] -> c -> (TVar Client -> IO x) -> IO ()
+withNats opts conn callback = do
+  -- apply initial opts and set puller
+  let client = newClient opts
+
+  let inbox = Buffer.newBuffer (recv conn)
+  let outbox =  []
+  let sender = send conn
+  let client' = client {inbox = inbox, outbox = outbox, sender = sender, closer = close conn}
+  tClient <- newTVarIO client'
+
+  waitForInfo tClient
+  routeMessages tClient
+
+  connect tClient
+  sendBytes tClient
+
+  callback tClient -- TODO: this will need to be async so we get to loop, but we need to know when it's done... or we make loop async, but keep track of it to cancel on close.. that would be nicer
+  sendBytes tClient
+
+  loop tClient -- TODO: ideally we want two loops, one for reading and one for writing
+
+loop :: TVar Client -> IO ()
+loop c = do
+  c' <- readTVarIO c
+  a <- newEmptyTMVarIO
+  b <- newEmptyTMVarIO
+  -- read incoming messages
+  void . forkIO $ do
+    case status c' of
+      CLOSING -> routeMessages c
+      _       -> readMessages c >> routeMessages c
+    atomically (putTMVar a ())
+  -- send outgoing messages
+  void $ forkIO (sendBytes c >> atomically (putTMVar b ()))
+  --wait for them to resolve
+  atomically $ takeTMVar a
+  atomically $ takeTMVar b
+  --decide what to do next
+  client <- readTVarIO c
+  case status client of
+    CLOSING -> cleanup c
+    _       -> loop c
+
+cleanup :: TVar Client -> IO ()
+cleanup client = do
+  client' <- readTVarIO client
+  -- if we've got nothing left to send or route then close the conn, otherwise send/route first then close
+  case flushed client' of
+    True -> closer client'
+    _    -> loop client
+
+  where
+    flushed c = out c && in' c
+    out c = null (outbox c)
+    in' c = (BS.length . Buffer.bytes . inbox $ c) == 0
+
+class NatsConn a where
+  recv :: a -> Int -> IO BS.ByteString
+  send :: a -> BS.ByteString -> IO ()
+  close :: a -> IO ()
+
+readMessages :: TVar Client -> IO ()
+readMessages client = do
+  client' <- readTVarIO client
+  inbox' <- execStateT Buffer.hydrate (inbox client')
```

Comment:
This is still questionable, it feeds a stale version of client into the state functions

Reply:
Arguably, if there is only one thread dealing with the inbox then it _shouldn't_ cause problems, but it might be worth adding a TMVar to illustrate this

Reply:
It might be that the inbox doesn't suit STM, and we should just pull directly in the client and atomically append that to a list of bytes, the only smarts the outbox has is tracking the buffer limit, which could arguably be done better in STM